### DISCUSSION TITLE
Simple Levels Stretch Help

### DISCUSSION THREAD

**Discussion by borneo1910** [ImageMagick/ImageMagick#4965] [Help]
Hello.  

I'm trying to write a script that replicates a fairly simple levels stretch in Photoshop.  I can approximate the results I want using -contrast-stretch 2%, but its not the same.  Here's what I'm trying to achieve.  

Given a histogram of a grey/flay field like so
![image](https://user-images.githubusercontent.com/71579476/158876463-8a146feb-472a-47e1-8236-7942d6172b96.png)

I want to stretch each color channel like this:  
![image](https://user-images.githubusercontent.com/71579476/158876924-f0a49378-c899-4ce0-8149-4ae292e1b980.png)
![image](https://user-images.githubusercontent.com/71579476/158876946-4a7e7bf9-b6fe-4206-91c0-2a48e440046c.png)
![image](https://user-images.githubusercontent.com/71579476/158876971-029609e5-3da8-48ca-b6a6-c01d2d25b53d.png)

To achieve a this:
![image](https://user-images.githubusercontent.com/71579476/158877061-0980d3c6-abd3-4c34-9b60-a6f68acf0cbb.png)

Then I would do it a second time (per channel)
![image](https://user-images.githubusercontent.com/71579476/158877344-f2452a26-b9fc-4477-b888-26c0c0b62605.png)
![image](https://user-images.githubusercontent.com/71579476/158877376-1764e7c0-815a-49da-ae54-6a2f4f7cb5b2.png)
![image](https://user-images.githubusercontent.com/71579476/158877405-56782ab4-b3f8-495c-a503-c8dddf46c8f9.png)

I'm also adjusting the grey point from about 1-.5 

I've tried multiple iterations of -level-colors but it only seems to perform the first and then nothing after that changes the image.  I'm sure I'm missing something.  I'm also not very experienced with IM.  

Any help would be greatly appreciated.

Thanks in advance
Mike

**Comment by fmw42** [comment]
You are looking for -level blackpoint,whitepoint,gamma. See https://imagemagick.org/script/command-line-options.php#level

However, you will need to convert your PS values to percent (by dividing by 255 and multiplying by 100). (ImageMagick values are in range 0 to 65535 or 0 to 100 percent for Q16 compile, not 0 to 255

Then use

`-level blackpoint%,whitepoint%,gamma`

The gamma is the middle value in the PS levels range. The blackpoint is the lower value on the left and whitepoint is the higher value on the right.

**Comment by borneo1910** [comment]
Thank you.

So just to be clear, do I use the option "blackout30%" explicityl?  ie I
need the words "blackoutXX$" and "whitepoint" etc?

On Thu, Mar 17, 2022 at 12:29 PM Fred Weinhaus ***@***.***>
wrote:

> You are looking for -level blackout,whitepoint,gamma. See
> https://imagemagick.org/script/command-line-options.php#level
>
> However, you will need to convert your PS values to percent (by dividing
> by 255 and multiplying by 100). (ImageMagick values are in range 0 to 65535
> or 0 to 100 percent for Q16 compile, not 0 to 255
>
> Then use
>
> -level blackout%,whitepoint%,gamma
>
> The gamma is the middle value in the PS levels range. The black point is
> the lower value on the left and white point is the higher value on the
> right.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ImageMagick/ImageMagick/discussions/4965#discussioncomment-2383097>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARCDOVAIBW22YJ4CVXL7SPDVAOBY5ANCNFSM5Q7Y4XEA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>


-- 
    Michael Borenstein, DIT
   818.414.2511
   <http://www.highlineimaging.com>

**Reply by fmw42** [reply]
Sorry, my compute auto spelling keeps changing what I type and I did not notice it. I have edited my post above to reflect the use of blackpoint and whitepoint.

**Reply by borneo1910** [reply]
No worries.  figured that part out via trial and error. 

what about using -level-colors?  Isn't that what I want to apply the level operator to the individual color channels?  When I try that route it doesn't do anything tho?

**Reply by fmw42** [reply]
No -level-colors is something completely different. It is for colorizing a grayscale image with a gradient of colors. 

If you want to do the PS levels on different color channels in ImageMagick, you could do them one at a time by selecting that channel. See https://legacy.imagemagick.org/Usage/color_mods/#level

`convert image -channel r -level .... +channel -channel g -level ... +channel -channel b -level +channel result`

**Comment by borneo1910** [comment]
Thank you!  That's what I needed to know!!

On Thu, Mar 17, 2022 at 2:10 PM Fred Weinhaus ***@***.***>
wrote:

> No -level-colors is something completely different. It is for colorizing a
> grayscale image with a gradient of colors.
>
> If you want to do the PS levels on different color channels in
> ImageMagick, you could do them one at a time by selecting that channel. See
> https://legacy.imagemagick.org/Usage/color_mods/#level
>
> convert image -channel r -level .... +channel -channel g -level ...
> +channel -channel b -level +channel result
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ImageMagick/ImageMagick/discussions/4965#discussioncomment-2383591>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARCDOVAQIXB2N5LCE6JLYJ3VAONVHANCNFSM5Q7Y4XEA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>


-- 
    Michael Borenstein, DIT
   818.414.2511
   <http://www.highlineimaging.com>

**Comment by borneo1910** [comment]
No worries.  figured that part out via trial and error.

what about using -level-colors?  Isn't that what I want to apply the level
operator to the individual color channels?  When I try that route it
doesn't do anything tho?

On Thu, Mar 17, 2022 at 1:01 PM Fred Weinhaus ***@***.***>
wrote:

> Sorry, my compute auto spelling keeps changing what I type and I did not
> notice it. I have edited my post above to reflect the use of blackpoint and
> whitepoint.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ImageMagick/ImageMagick/discussions/4965#discussioncomment-2383265>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARCDOVCO7WW6BZPP3BKWVZLVAOFR5ANCNFSM5Q7Y4XEA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>


-- 
    Michael Borenstein, DIT
   818.414.2511
   <http://www.highlineimaging.com>

**Reply by fmw42** [reply]
No.  The operator is +level-colors and it is used to recolor grayscale images. -level-colors maps colors to black and white (and is generally used only limitedly).  See https://imagemagick.org/script/command-line-options.php#level-colors
